### ISSUE TITLE
4.04 --master-display not showing in media info due to SEI_rbsp_stop_one_bit Missing

### ISSUE THREAD

**Issue by pwacooijmans** [rigaya/NVEnc#51]
I used this command:
NVEncC64 --sar 1:1 --codec h265 --preset quality --output-depth 10 --chromaloc 2 --videoformat pal --strict-gop --vbr 25000 --max-bitrate 40000 --vbv-bufsize 160000 --master-display G(13250,34500)B(7500,3000)R(34000,16000)WP(15635,16450)L(10000000,0) --max-cll 1000,640 --colormatrix bt2020nc --colorprim bt2020 --transfer smpte2084 --mv-precision Q-pel -i "D:\La La Land (2016)\La La Land (2016)_temp\La La Land (2016)_new.avs" -o "D:\La La Land (2016)\La La Land (2016)_temp\La La Land (2016)_new_out.h265"

and this is the result in media info as you see no master display info:
![no master display info](https://user-images.githubusercontent.com/30800180/40619317-e1bbaa40-6294-11e8-821c-1c8782a8683c.jpg)

it should say something like this:
![with master display info](https://user-images.githubusercontent.com/30800180/40619443-564425e0-6295-11e8-9bc0-b837ca39f063.jpg)

**Comment by pwacooijmans** [comment]
in other words the amounts you enter as master display values are like this:
G(x*50000,y*50000)B(x*50000,y*50000)R(x*50000,y*50000)WP(x*50000,y*50000)L(x*10000,y*10000)

**Comment by pwacooijmans** [comment]
https://github.com/SK-Hardwired/nv_hevc_hdr_patcher
just check this link if you need help on getting it to work :)

**Comment by pwacooijmans** [comment]
This is what the maker of nvhsp wrote to me in an e-mail....and I want you to have it

DCI-P3:	--master-display G(13250,34500)B(7500,3000)R(34000,16000)WP(15635,16450)L(10000000,1) --max-cll 1000,1
RGB:	G(x=0.265, y=0.690), B(x=0.150, y=0.060), R(x=0.680, y=0.320), WP(x=0.3127, y=0.329), L(max=1000, min=0.0000)

Rec.709:	--master-display G(15000,30000)B(7500,3000)R(32000,16500)WP(15635,16450)L(10000000,1) --max-cll 1000,1
RGB:	G(x=0.30, y=0.60), B(x=0.150, y=0.060), R(x=0.640, y=0.330), WP(x=0.3127,y=0.329), L(max=1000, min=0.0000)

Rec.2020:	--master-display G(8500,39850)B(6550,2300)R(35400,14600)WP(15635,16450)L(10000000,1) --max-cll 1000,1
RGB:	G(x=0.170, y=0.797), B(x=0.131, y=0.046), R(x=0.708, y=0.292), WP(x=0.3127,y=0.329), L(max=1000, min=0.0000)

To convert, simply divide the RGB values by 0.00002
Sample for Display P3:
0.265 / 0.00002 = 13250
0.69 / 0.00002 = 34500
0.15 / 0.00002 = 7500
0.06 / 0.00002 = 3000
0.68 / 0.00002 = 34000
0.32 / 0.00002 = 16000
0.3127 / 0.00002 = 15635
0.329 / 0.00002 = 16450

Information:
https://en.wikipedia.org/wiki/DCI-P3
https://en.wikipedia.org/wiki/Rec._709
https://en.wikipedia.org/wiki/Rec._2020

https://www.smpte.org/sites/default/files/section-files/HDR.pdf

**Comment by pwacooijmans** [comment]
How do you do the math that results in the result? Is there a formula for that?

 @drSHLEFF
drSHLEFF commented on 22 Dec 2017
Yes
--master-display 
SMPTE ST 2086 mastering display color volume SEI info, specified as a string which is parsed when the stream header SEI are emitted. The string format is “G(%hu,%hu)B(%hu,%hu)R(%hu,%hu)WP(%hu,%hu)L(%u,%u)” where %hu are unsigned 16bit integers and %u are unsigned 32bit integers. The SEI includes X,Y display primaries for RGB channels and white point (WP) in units of 0.00002 and max,min luminance (L) values in units of 0.0001 candela per meter square. Applicable for HDR content.

Example for a P3D65 1000-nits monitor, where G(x=0.265, y=0.690), B(x=0.150, y=0.060), R(x=0.680, y=0.320), WP(x=0.3127, y=0.3290), L(max=1000, min=0.0001):

G(13250,34500)B(7500,3000)R(34000,16000)WP(15635,16450)L(10000000,1)
Note that this string value will need to be escaped or quoted to protect against shell expansion on many platforms. No default.

--max-cll 
Maximum content light level (MaxCLL) and maximum frame average light level (MaxFALL) as required by the Consumer Electronics Association 861.3 specification.

Specified as a string which is parsed when the stream header SEI are emitted. The string format is “%hu,%hu” where %hu are unsigned 16bit integers. The first value is the max content light level (or 0 if no maximum is indicated), the second value is the maximum picture average light level (or 0). Applicable for HDR content.

Example for MaxCLL=1000 candela per square meter, MaxFALL=400 candela per square meter:

–max-cll ?1000,400?
Note that this string value will need to be escaped or quoted to protect against shell expansion on many platforms. No default.

http://x265.readthedocs.io/en/default/cli.html

**Comment by drSHLEFF** [comment]
I'm not the creator of nvhsp, I just saw your request ...)

**Comment by pwacooijmans** [comment]
Thx anyway :)

**Comment by pwacooijmans** [comment]
Tried it again with this line WITH quotes this time....still the same issue:
NVEncC64 --sar %target_sar% --codec h265 --preset quality --output-depth 10 --chromaloc 2 --videoformat pal --strict-gop --vbr 25000 --max-bitrate 40000 --vbv-bufsize 160000 --master-display `G(13250,34500)B(7500,3000)R(34000,16000)WP(15635,16450)L(10000000,0)` --max-cll '1000,640' --colormatrix bt2020nc --colorprim bt2020 --transfer smpte2084 --mv-precision Q-pel -i "%script_file%" -o "%encoder_out_file%"

**Comment by pwacooijmans** [comment]
Thank you SO much for this tool....extremely handy :)

Many thx for all your hard work....you can close the ticket

2018-05-29 16:10 GMT+02:00 rigaya <notifications@github.com>:

**Comment by JeromeMartinez** [comment]
>  even though MediaInfo will not show the information

Is it possible to get a file with no mastering display information in MediaInfo despite its presence in the stream? Thanks.

**Comment by pwacooijmans** [comment]
Depends on the tool you read it with...some are better than others

Op wo 30 mei 2018 10:11 schreef Jérôme Martinez <notifications@github.com>:

**Comment by JeromeMartinez** [comment]
> Depends on the tool you read it with...some are better than others

This is the weird part: the spec is pretty basic, so it does not make sense here to have "some are better than others".
It would help at a low cost for you to improve tools if just a sample file is provided: 1 of the tools is buggy here, and if I get a file I could check the bitstream against the spec (and if issue in on MediaInfo side, I'll fix it).
But without a file, implying that a specific tool is buggy without permitting its author to check harms rather than helps.
Please provide a sample file demonstrating the issue.

**Comment by pwacooijmans** [comment]
Furthermore I noticed if you have a minimum L value in the masterdisplay
string that is equal to 0....change it to 1.....0 happens to be undefined
in nvenc

Op wo 30 mei 2018 10:22 schreef Pieter Cooijmans <pwacooijmans@gmail.com>:

**Comment by pwacooijmans** [comment]
Deal..you'll get your 1 frame with 4.05

Op wo 30 mei 2018 10:50 schreef Jérôme Martinez <notifications@github.com>:

**Comment by pwacooijmans** [comment]
here a google drive link with a sample file
https://drive.google.com/open?id=1_hE55gIqvMX5a5VuhA0BG4HF-XTjNXqX

2018-05-30 10:53 GMT+02:00 Pieter Cooijmans <pwacooijmans@gmail.com>:

**Comment by JeromeMartinez** [comment]
> here a google drive link with a sample file

Thanks.
Looks like MediaInfo shows everything with this file, at least with latest version of MediaInfo:

```
Color primaries                          : BT.2020
Transfer characteristics                 : PQ
Matrix coefficients                      : BT.2020 non-constant
Mastering display color primaries        : Display P3
Mastering display luminance              : min: 0.0001 cd/m2, max: 1000 cd/m2
Maximum Content Light Level              : 1000 cd/m2
Maximum Frame-Average Light Level        : 640 cd/m2
```

(the numbers you set on the command line are understood as "Display P3" values, so numbers are not directly showed)

So IMO everything actually looks fine everywhere (match between NVEnc command and MediaInfo analysis, and created stream conforms to specs the same way x265 does), at least with latest versions of tools.

**Comment by pwacooijmans** [comment]
It's just that if you enter random number it won't work..from time to
time.....thats why I only use dci p3, rec 2020 and rec709 to be more
compatible with mediaplayers/tvs in general

Op wo 30 mei 2018 11:40 schreef Jérôme Martinez <notifications@github.com>:

**Comment by JeromeMartinez** [comment]
> It's just that if you enter random number it won't work

Please provide a sample file "it won't work". I expect that MediaInfo shows "raw" values in that case, not hiding them from the user.
Actually what is needed is the example used [when you initiated the issue](https://github.com/rigaya/NVEnc/issues/51#issue-327047293) that leads to a "even though MediaInfo will not show the information" comment.

> thats why I only use dci p3, rec 2020 and rec709 to be more compatible with mediaplayers/tvs in general

True, but this is only a limitation of the mediaplayers/tvs, and not linked to any encoder "method" for stroring values.

**Comment by pwacooijmans** [comment]
Ok will try

Op wo 30 mei 2018 11:53 schreef Jérôme Martinez <notifications@github.com>:

**Comment by pwacooijmans** [comment]
On the same google drive link I send you is now a sample of a file that
fails with the media info in the mediaplayer: mpc-be x64

Op wo 30 mei 2018 11:54 schreef Pieter Cooijmans <pwacooijmans@gmail.com>:

**Comment by pwacooijmans** [comment]
Oh and by the way the error 15 means you have NOT installed nvidia driver
397.93

Op wo 30 mei 2018 12:25 schreef Pieter Cooijmans <pwacooijmans@gmail.com>:

**Comment by JeromeMartinez** [comment]
> On the same google drive link I send you is now a sample of a file that fails with the media info

You set min_display_mastering_luminance to 0, right? (edit: OK, I see that in the screenshot)
This leads to 0 in the last byte of the SEI.
Who is encoding this SEI?

Checking both files you provided, I see that "rbsp_trailing_bits()" bit is missing from SEI packets.
In MediaInfo, we try to conform to spec, so we discard extra zeroes at the end, from specs:

```
NumBytesInNalUnit is set equal to the number of bytes starting with the byte at the current position in the byte
stream up to and including the last byte that precedes the location of one or more of the following conditions:
– A subsequent byte-aligned three-byte sequence equal to 0x000000,
– A subsequent byte-aligned three-byte sequence equal to 0x000001,
– The end of the byte stream, as determined by unspecified means.
```

Issue is the 1st part, and MediaInfo is conforming to it.

With content from your file starting at offset 0x80:
00 00 03 **_00 00_** 00 00 00 01

Bytes 0x83 and 0x84 (in bold) are not integrated in NumBytesInNalUnit because the sequence is 0x000000, despite they are part of it from the encoder point of view.
So 2 bytes are missing on the SEI, leading to discard the SEI by MediaInfo because the SEI is not complete.
IMO  hevcesbrowser is too much tolerant about trailing zeroes, and is not compliant (in other cases, it would misinterpret a stream with extra zeroes on purpose).
Note: for the first file, MediaInfo detects the missing rbsp_trailing_bits( ) only after having parsed min_display_mastering_luminance, so the issue is less seen, you can see that there is an issue only if you use MediaInfo in a conformance checker mode.

From spec:
```
sei_rbsp( ) {
    do
        sei_message( )
    while( more_rbsp_data( ) )
    rbsp_trailing_bits( )
}
```

I don't see "rbsp_trailing_bits( )" part, i.e. a pending 1-bit marker set to 1 in the SEI (in that case, as it is byte aligned, a byte set to value 0x80 is missing in the SEI).

IMO the stream is invalid and may lead to have the SEI discarded by some decoders, with reasons (invalid SEI).

@rigaya IMO output by NVEnc 4.05 has a general problem, it forgets "rbsp_trailing_bits( )" at the end of the SEI, which leads to an invalid stream, and conforming players may discard the SEI (maybe some decoders are more tolerant, but they violate the spec). It would be safer to have a output a stream conforming to spec, and "even though MediaInfo will not show the information" is a messenger saying that something is wrong somewhere.

I may be wrong, I would be happy to know it with arguments (how the SEI stored in this stream is conforming to specs) in order to fix MediaInfo for conforming to spec, if the stream is conforming to spec, but all files I have (also not created by x265) has this extra byte with value 0x80 at the end of each SEI.

> Oh and the thing is x265 CAN repeat NALS and nvenc I don't think that is possible

repetition is not mandatory, not the issue here IMO.

> Oh and by the way the error 15 means you have NOT installed nvidia driver 397.93

Hum... True, I missed this prerequisite, I used the one installed on my machine (not up to date), good to know for next time (error message is not obvious...).

**Comment by pwacooijmans** [comment]
Extremely much thanks for getting this detailed into the data...hopefully
rigaya can fix it :)

Op wo 30 mei 2018 13:43 schreef Jérôme Martinez <notifications@github.com>:

**Comment by pwacooijmans** [comment]
Well you said it yourself now....no one sticks to the simple set of rules....it's just my experience as a programmer sorry :)....communication and making people understand how important sticking to a standard is is paramount! So I really appreciate your effort in trying to do so. Since you are probably one of the few people who KNOWS the standard and tries to stick to it :)

**Comment by pwacooijmans** [comment]
By the way doesn't this mean mediainfo isn't looking at the trailing bit correct either in case it's exactly a standard like dci-p3,rec 2020 and rec 709?.....It's just a thought, hope no offense is taken.

**Comment by pwacooijmans** [comment]
But how come when the dci-p3 norm is set with L(1000,1) it does show in
media info?

Op do 31 mei 2018 11:14 schreef Jérôme Martinez <notifications@github.com>:

**Comment by pwacooijmans** [comment]
It WILL show dci-p3 in that case...even though the trailing bit is missing

Op do 31 mei 2018 11:18 schreef Pieter Cooijmans <pwacooijmans@gmail.com>:

**Comment by JeromeMartinez** [comment]
> is it possible to disable showing ANY malformed SEI INCLUDING the dci-p3,rec 2020,rec 709 cuz this is damn confusing :)

It would not be confusing if stream is valid. confusion is due to encoder not encoding correctly the SEI block.
And current behavior is conform to spec ("unspecified" when stream is not compliant).
No plan to change such behavior for the moment (development cost for very low number of cases, and no sponsor interested in paying the development, and little interest for that as issue is on encoder side).

**Comment by pwacooijmans** [comment]
ah well pitty.....if I had the money you'd get it but oh well just an ordinary worker bee :)

**Comment by pwacooijmans** [comment]
Just do me the favour of writing down the issue....who knows someday someone will come along and fix it :)

**Comment by JeromeMartinez** [comment]
I understand... But I was interested in fixing any MediaInfo bug, and for the moment your request is a feature request, and I already have tons of them in my free support list.

But really, MediaInfo is here just a messenger, the only sustainable fix is to have the encoder encode correctly the SEI (easy fix, and better to be compliant, isn't it?), and currently we have no response about that.

> someone will come along and fix it

This is not a fix, but a feature implementation, as current behavior of MediaInfo is conform to spec.

**Comment by pwacooijmans** [comment]
give the Rigaya some time, who knows he might even be doing this in his free time as a hobby :)

**Comment by pwacooijmans** [comment]
Well last attempt to make a feature request for media info :) Could you add 1 if statement that checks IF the minimum luminance=1 AND the trailing bit is present to show dci-p3,rec 2020,rec 709 otherwise do it normally conform to your spec? :)

**Comment by pwacooijmans** [comment]
pretty please with a cherry on top :)

**Comment by rigaya** [comment]
According to your conversation, I think changing NVEnc output will do.

I'll change the output when I have time, and test again.

**Comment by pwacooijmans** [comment]
thx a LOT rigaya :)

**Comment by JeromeMartinez** [comment]
> Well last attempt to make a feature request for media info :) Could you add 1 if statement that checks IF the minimum luminance=1 AND the trailing bit is present to show dci-p3,rec 2020,rec 709

it does not make sense (why "1" and not any other value?).
you don't catch that it does not depend on any value, issue is more global. Behavior may be different (e.g. 0 value is displayed if there is another SEI message after this SEI message)
Current code does not catch presence or lack of trailing bit at this place, need to change stuff (parse the complete SEI before filling data from a specific message).
that means that such change requests to study well the exact result for lot of cases. I don't see any reason to risk regressions detected by my sponsors.

> your spec? 

I have no "my spec". I have H.265 spec and my code. 

> pretty please with a cherry on top :)

The only right thing to do is to fix the encoder, really. No need about something else.

**Comment by pwacooijmans** [comment]
Sorry for persisting, it was a joke :) I just thought ANY wrong SEI would a HELL of a lot of work :)

**Comment by pwacooijmans** [comment]
ANY as in ANY future versions too :).....bloody any key can never find it :)

**Comment by pwacooijmans** [comment]
It's just we both know this can be prevented with 1 if then else statement saying: if sei complete AND the trailbit exists then post it else ignore and report bug

**Comment by pwacooijmans** [comment]
I just tell my team of programmers one thing: will you get in your own coded self driving car alive and drive it knowing your kid put a chewinggum on the camera

**Comment by pwacooijmans** [comment]
we even keep a pedestrian kill count....I am not the one with the least kills but a close second :)

**Comment by pwacooijmans** [comment]
I know we have different visions about programming and stuff but this is the way I look at it: IF you invent a drivers license and but you do NOT make a law to prevent people from not sticking to that drivers lisence...what will happen?

**Comment by pwacooijmans** [comment]
in the end no one will buy the drivers lisence....a voila out of work

**Comment by pwacooijmans** [comment]
cuz now I know mediainfo is not checking for the dci-p3 rec2020 and rec709 AND the trailing bit it just lost it's value for me and I am forced to use hevcesbrowser instead....it's just a fact....no offense intended

**Comment by JeromeMartinez** [comment]
> and I am forced to use hevcesbrowser instead

And you don't know if your TV will handle correctly the metadata (it actually depends on the TV model, how invalid SEI are handled). This is just a bug in the encoder so you get incoherent results, it will be fixed on encoder side, that's all.
On my side I was just wanting a bug if on MediaInfo side, it is not on MediaInfo side, I helped about the reason there is a bug in the encoder, that's all. No plan to debate more about how to handle invalid streams.

**Comment by pwacooijmans** [comment]
Well you know I spend thousands of hours encoding movies and I will never get them back just due to this little bug :(

**Comment by pwacooijmans** [comment]
since all of a sudden the masterdisplay strings on old movies stopped appearing

**Comment by pwacooijmans** [comment]
it's at least 1 year of work for 1 man to fix this all....so yeah you can understand I am not very happy :)

**Comment by JeromeMartinez** [comment]
> since all of a sudden the masterdisplay strings on old movies stopped appearing

Stopped where?

>  it just lost it's value for me and I am forced to use hevcesbrowser instead

Not sure I understand the issue on your side, but if you want support of buggy SEI in MediaInfo, please vote for [the corresponding feature request on MediaInfo website](https://mediaarea.net/Vote/Support-Buggy-Hevc-H-265-Sei-171). If there are 15 or more points (which is far lower tha the real cost for me), I will implement something for supporting the first file you provided (and indicated that the stream is buggy so maybe not parsed correctly everywhere)

**Comment by pwacooijmans** [comment]
a voila you got your 15 points :) now find some time to fix it :)

**Comment by pwacooijmans** [comment]
Thanks a lot to you both you just supported indirectly free instruction videos that work for people at home here in holland for cpr and things like that at a local basis :)

**Comment by pwacooijmans** [comment]
The thing you said about I don't know how tvs handle input, that's true. But you didn't see your lifetime work on a echo scanner that can look under the skin and detect cancer in an early stage WITHOUT radiation at a price SO low EVERY household can afford it based on cuda technology go to waste due to philips and hospitals not seeing the reason to invest in something this cheap. I have nda but I don't give a shit anymore sorry.

**Comment by pwacooijmans** [comment]
Wil do

Op za 2 jun. 2018 11:24 schreef Jérôme Martinez <notifications@github.com>:

**Comment by pwacooijmans** [comment]
https://drive.google.com/open?id=1_hE55gIqvMX5a5VuhA0BG4HF-XTjNXqX

2018-06-02 12:32 GMT+02:00 Pieter Cooijmans <pwacooijmans@gmail.com>:

**Comment by JeromeMartinez** [comment]
I confirm that the files encoded with 4.06 are better (final bit present).
Don't hesitate to check with other tools in order to be sure all tools handle well new streams.

PS: file names are inverted ;-), but it does not matter.

**Comment by pwacooijmans** [comment]
ah well sorry for that just wanted to get it done asap for you :)...since
you helped the community SOOO much :)

2018-06-02 12:39 GMT+02:00 Jérôme Martinez <notifications@github.com>:

**Comment by pwacooijmans** [comment]
Doing that as we speak....so far everything seems to check out fine

2018-06-02 12:40 GMT+02:00 Pieter Cooijmans <pwacooijmans@gmail.com>:

**Comment by pwacooijmans** [comment]
If you happen to find any obvious errors please let rigaya know it would be
highly appreciated :)

2018-06-02 12:45 GMT+02:00 Pieter Cooijmans <pwacooijmans@gmail.com>:

**Comment by pwacooijmans** [comment]
changed the file names

2018-06-02 12:46 GMT+02:00 Pieter Cooijmans <pwacooijmans@gmail.com>:

**Comment by Revan654** [comment]
Following Is output before all HDR Metadata has been inserted:

![output](https://user-images.githubusercontent.com/38091898/41072005-2ade22f8-69c9-11e8-86d1-f76ae7ad69c8.jpg)
![output mkv](https://user-images.githubusercontent.com/38091898/41072016-2fb785bc-69c9-11e8-83d2-70970c9e830c.jpg)

Now HDR with proper data and Color Coordinates.

![hdrmetadata info](https://user-images.githubusercontent.com/38091898/41072030-4a09fac6-69c9-11e8-95f6-b3b3883e014e.jpg)
![hdrmetadata](https://user-images.githubusercontent.com/38091898/41072033-4e1c2986-69c9-11e8-9ea8-130589ca57c2.jpg)

I'm not sure if mediainfo can display the color coordinates or not.

---------

With NVenc(Not to be Confused with NVEnCc) is missing a bunch of Flags, Which can be found on the software side(CPU encoding). Not sure if they have different name or not.

These are the Missing Flags for HDR10:
--hdr 
--repeat-headers 
--hrd 
--aud 
--vui-hrd-info
- --vbv-bufsize and --vbv-maxrate (Only one of these flags can be found in the encoder, unless I missed it).

I'll add this too:

it's common to master HDR content in the DCI P3 color space, with either the DCI (~D50) or D65 white points. **This is not a supported format for delivery to consumer electronics.** When mastering, choose Rec. 2020 color primaries (the Rec. 2100 standard implies Rec. 2020 color in many applications).
A common mistake is to master in P3, then tag the result using Rec. 2020 primaries. This will result in an oversaturated look with shifted hues.

**Comment by JeromeMartinez** [comment]
> I'm not sure if mediainfo can display the color coordinates or not.

Is there any interest? standard name and standard raw values are synonyms, I just preferred to show the name of the standard when values have a name, I think it is easier to read.

**Comment by pwacooijmans** [comment]
There has always been an interest on my side...however i spend all the
money i could spare...so can't invest right now

Op do 7 jun. 2018 15:32 schreef Jérôme Martinez <notifications@github.com>:

**Comment by pwacooijmans** [comment]
Who knows maybe next month sorry I am on a tight budget :)

**Comment by JeromeMartinez** [comment]
the question is not about budget right now, it is why you prefer to check manually 8 numbers rather than a name for knowing something.
I could understand for some automation... Actually not, with automation it is easy to convert a name to 8 numbers.
So for the moment no plan to change a name to 8 complicated numbers in the output.
(and why now? in the last 10 years, it was not requested to show 8 numbers for the colors primaries, as well as raw numbers for transfer characteristics and matrix coefficients, it is weird that the request is only on 1 line of MediaInfo, and not all the similar lines)

**Comment by pwacooijmans** [comment]
well for a starter not everyone knows the dci-p3, rec2020, rec709 numbers by head....some don't even have a clue. So when you set it to dci-p3 it makes it very hard for people to guess what the correct numbers are.

**Comment by pwacooijmans** [comment]
And it's not like someone doesn't want to make their own videos with hdr or something

**Comment by pwacooijmans** [comment]
So mediainfo as a teaching tool is very handy from time to time

**Comment by JeromeMartinez** [comment]
> well for a starter not everyone knows the dci-p3, rec2020, rec709 numbers by head....

Argument for how MediaInfo currently shows it, not against. so if you think names are better, why do you want MediaInfo to display numbers? Still no argument for (only against).

**Comment by pwacooijmans** [comment]
it's not really againt....I will give you an example:
https://en.wikipedia.org/wiki/DCI-P3
People usually try to guess the right numbers from this little picture on the right of the website....try to do it that precise that you can get the right values....it's freaking hard.

**Comment by pwacooijmans** [comment]
Since they want to stick to dci-p3 too....so when they have an example movie that should dci-p3 and they KNOW it is dci-p3 they can't replicate it due to the numbers not showing.

**Comment by pwacooijmans** [comment]
It's not about knowing the dci-p3 standards and such...it's just about the correct input

**Comment by JeromeMartinez** [comment]
> People usually try to guess the right numbers from this little picture on the right of the website....

???
Exact numbers are https://en.wikipedia.org/wiki/DCI-P3#System_colorimetry
tools could support names instead of values (as it is done for "--colormatrix bt2020nc --colorprim bt2020 --transfer smpte208"), input method does not depend on how it is stored in the file, for easiness.

**Comment by pwacooijmans** [comment]
yeah duh those are obvious....but that doesn't mean they KNOW the standard...they just watched one dude say on internet you NEED to stick to this bla bla standard...they copy paste it e voila...same problem

**Comment by pwacooijmans** [comment]
That's what people run in to on a daily basis.

**Comment by pwacooijmans** [comment]
if I say to you when you don't know english 1 million people died today......no one has a clue.....if I say 1000000 dead....they get it....catch my drift

**Comment by pwacooijmans** [comment]
math is the only universal language there is....be happy with it :)

**Comment by pwacooijmans** [comment]
and if I say to you big news: 1 million people died today....they just copy paste it without having a clue of the true meaning

**Comment by pwacooijmans** [comment]
and then you get people that just copy paste dci-p3 instead of the actual numbers required for encoding e voila error.....and they go stupid program....etc

**Comment by pwacooijmans** [comment]
need any more examples...cuz I can keep this going for 10 hours straight if you like

**Comment by pwacooijmans** [comment]
in short I like your idea...however no encoding program supports it and that's the issue here.

**Comment by pwacooijmans** [comment]
Now I know mediainfo sticks to the actual scientific output of the master display string...but if you REALLY want to help the public...just convert them to integers as used in the master display string.

**Comment by pwacooijmans** [comment]
and keep it in the order of the master display string....in fact...create a master display string they only have to copy and paste.

**Comment by pwacooijmans** [comment]
that would up your program through the roof...just a hint

**Comment by pwacooijmans** [comment]
it's 10000 times faster than having to manually calculate G(X times 50000,Y times 50000)B(X Times 50000,Y Times 50000)R(X Times 50000, Y Times 50000)L(X Times 10000, Y Times 10000) and it's more bug and idiot proof

**Comment by pwacooijmans** [comment]
and just call it master display string instead
as always keep it as simple as possible

**Comment by pwacooijmans** [comment]
the best media info there is would be one that you copy 1 line of code and you get the exact settings needed for encoding...however that can't happen since encoders use different standards.

**Comment by pwacooijmans** [comment]
but just keep that in mind....as a guide to build your program

**Comment by JeromeMartinez** [comment]
> create a master display string they only have to copy and paste.

Your usage is not the one of another person.
And most people are interested in knowing the standard name, not having values (they have to check 8 numbers for knowing if this is what they look for)
having 8 numbers is useless for most people (especially because it will rarely be the format expected by all users at the same time, as format of the line depends on the tool...)
If you want a "NVEnc profile" (i.e. provide the command line for creatingt he same output) in MediaInfo, doable with some sponsoring.
But please stop to imagine that your usage is the one of all other people on earth, it does not help in the discussion.

**Comment by pwacooijmans** [comment]
Just sharing my view, sorry if you feel it that way. However x265 also uses this format it's not just for nvenc....but yeah the sponsoring part is the issue here for me.

**Comment by pwacooijmans** [comment]
But if you have so many people buying and investing in your product that twitch.tv streams and youtube streams don't matter anymore.....then I am totally wasting your time and sorry for that.

**Comment by Revan654** [comment]
> Is there any interest? standard name and standard raw values are synonyms, I just preferred to show the name of the standard when values have a name, I think it is easier to read.

I would say yes, The main reason would be to see if the metadata is correct for Upload on Youtube if your planning to upload HDR content. Your metadata has to be formatted a certain way for upload or site doesn't fully accept it. Google then goes ahead and applies there own form of 3DLut or creates own Color Coordinates. When can create a horrible look.

> well for a starter not everyone knows the dci-p3, rec2020, rec709 numbers by head....some don't even have a clue. So when you set it to dci-p3 it makes it very hard for people to guess what the correct numbers are.

You shouldn't be guessing to start with. The Master-Display Numbers come directly from the monitor you color graded on. Of course if your making a local copy of a 4KHDR movie it's impossible to know the actual monitor that was used.

That Data would come from Davinci Resolve or similar software.

---------

There are different forums of HDR encoding by NVENC. This data comes directly from a custom made encoder (Not NVencC). Which works hand and hand with the company's HDR capture cards.

![directhevc](https://user-images.githubusercontent.com/38091898/41119398-b4279ab6-6a60-11e8-8aef-7c3f684b9219.jpg)

As you can see nothing about HDR is seen, Even though the file is HDR and has all HDMI InfoFrame data from the HDR source.

I just don't think there is a way for Mediainfo to cover all different forms of HDR.

This is uncompressed HDR file(which doesn't have any color grading or HDR metadata applied yet):

![rawhdr](https://user-images.githubusercontent.com/38091898/41119534-07f68da0-6a61-11e8-91fd-b5e38271284c.jpg)

^ The hardware needed to pull this off is very expensive. Since you need SSD in Raid0 and enough space for the actual capture file.

12 Seconds uncompressed is 8 gigs.

**Comment by JeromeMartinez** [comment]
> The main reason would be to see if the metadata is correct for Upload on Youtube if your planning to upload HDR content. Your metadata has to be formatted a certain way 

IMO this is an argument **for** names, not raw values, as raw values are harder to read.
Easier to see that your file is conform by checking a name instead of checking a typo in 8 numbers, don't you think?

> As you can see nothing about HDR is seen, Even though the file is HDR and has all HDMI InfoFrame data from the HDR source.

Please provide a sample file, it is not normal that MediaInfo does not display HDR metadata if it is present.
But... Note that you see HDR info (mastering stuff), only the other fields (Min/Max values) are missing.

> I just don't think there is a way for Mediainfo to cover all different forms of HDR.

When raw values are not a known standard, raw values are displayed. So all is covered, at least with the metadata format supported by MediaInfo.

> This is uncompressed HDR file(

I see I don't map P010 to YUV 4:2:0 , I'll do it soon.

> 12 Seconds uncompressed is 8 gigs.

"Only", because you have subsampling and 10-bit.
If you want to play, I sometimes work with overscanned 4K without subsampling (RGB) and at 16 bit / component, so each frame has a size of 120 MB, this leads to a bitrate of 50 Gbps for 60 fps stuff, 10x the bitrate of your file ;-)

**Comment by pwacooijmans** [comment]
If I am correct the new game of thrones 4K series has a slightly different one.....though I don't own it. So sorry can't help.

**Comment by Revan654** [comment]
> Now you are just showing off :) pimp my computer with that hardware will you :) 

If you saw my PC, you would be amazed. I also built a custom waterloop for it.

> "Only", because you have subsampling and 10-bit.
If you want to play, I sometimes work with overscanned 4K without subsampling (RGB) and at 16 bit / component, so each frame has a size of 120 MB, this leads to a bitrate of 50 Gbps for 60 fps stuff, 10x the bitrate of your file ;-)

This is directly from a capture card, There are limitations due to HDMI standards. The card I have can do RGB without any problem. It's HDMI that has problems accept it.

> Please provide a sample file, it is not normal that MediaInfo does not display HDR metadata if it is present.
But... Note that you see HDR info (mastering stuff), only the other fields (Min/Max values) are missing.

Tell me where to upload it to and I will send a sample. I have to review the source code to see what Flags are being enabled. Since the software was just released publicly. 

I know the file has HDMI-HDR metadata. I still have to look up what the difference is between HDMI HDR Info Frame data and other HDR metadata.

**Comment by JeromeMartinez** [comment]
> Tell me where to upload it to and I will send a sample.

Anywhere (any file upload website would be fine). If you need a private FTP access, please contact us at info@mediaarea.net.

**Comment by rigaya** [comment]
I'll close this issue as it's finally fixed in [NVEnc 4.55](https://github.com/rigaya/NVEnc/releases/tag/4.55) when muxing into containers, please look at #75 for further information.
